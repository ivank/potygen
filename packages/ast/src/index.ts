import { Parser } from '@ikerin/rd-parse';
import { Grammar } from './grammar';
import { QueryTag, WithTag } from './grammar.types';

export {
  AnyCastTag,
  AnyTypeTag,
  ArrayConstructorTag,
  ArrayIndexRangeTag,
  ArrayIndexTag,
  AsTag,
  BetweenTag,
  BinaryExpressionTag,
  BinaryOperatorTag,
  BooleanTag,
  CaseTag,
  CastableDataTypeTag,
  CastTag,
  CollateTag,
  ColumnsTag,
  ColumnTag,
  CombinationTag,
  ComparationExpressionTag,
  ComparationOperatorTag,
  ConflictConstraintTag,
  ConflictTag,
  ConflictTargetTag,
  ConstantTag,
  CountTag,
  CTETag,
  DataTypeTag,
  DefaultTag,
  DeleteTag,
  DistinctTag,
  DoNothingTag,
  DoUpdateTag,
  ElseTag,
  ExpressionListTag,
  ExpressionTag,
  FromListItemTag,
  FromListTag,
  FromTag,
  FunctionTag,
  GroupByTag,
  HavingTag,
  IdentifierTag,
  InsertTag,
  JoinOnTag,
  JoinTag,
  JoinTypeTag,
  JoinUsingTag,
  LimitTag,
  NullTag,
  NumberTag,
  OffsetTag,
  OperatorExpressionTag,
  OrderByItemTag,
  OrderByTag,
  OrderDirectionTag,
  ParameterTag,
  PgCastTag,
  QuotedNameTag,
  ReturningListItemTag,
  ReturningTag,
  RowTag,
  SelectListItemTag,
  SelectListTag,
  SelectParts,
  SelectTag,
  SetItemTag,
  SetListTag,
  SetMapTag,
  SetTag,
  SqlTag,
  StarIdentifierTag,
  StarTag,
  StringTag,
  TableTag,
  Tag,
  TypeArrayTag,
  TypeTag,
  UnaryExpressionTag,
  UnaryOperatorTag,
  UpdateFromTag,
  UpdateTag,
  UsingTag,
  ValuesListTag,
  ValuesTag,
  WhenTag,
  WhereTag,
  WithTag,
} from './grammar.types';

export {
  isAnyCast,
  isAnyType,
  isArrayConstructor,
  isArrayIndex,
  isArrayIndexRange,
  isAs,
  isBetween,
  isBinaryExpression,
  isBinaryOperator,
  isBoolean,
  isCase,
  isCast,
  isCastableDataType,
  isCollateTag,
  isColumn,
  isColumns,
  isCombination,
  isComparationExpression,
  isComparationOperator,
  isConflict,
  isConflictConstraint,
  isConflictTarget,
  isConstant,
  isCount,
  isCTETag,
  isDataType,
  isDefault,
  isDelete,
  isDistinct,
  isDoNothing,
  isDoUpdate,
  isElse,
  isExpression,
  isExpressionList,
  isFrom,
  isFromList,
  isFromListItem,
  isFunction,
  isGroupBy,
  isHaving,
  isIdentifier,
  isInsertTag,
  isJoin,
  isJoinOn,
  isJoinType,
  isJoinUsing,
  isLimit,
  isNull,
  isNumber,
  isOffset,
  isOperatorExpression,
  isOrderBy,
  isOrderByItem,
  isOrderDirection,
  isParameter,
  isPgCast,
  isQuotedName,
  isReturning,
  isReturningListItem,
  isRow,
  isSelect,
  isSelectList,
  isSelectListItem,
  isSelectTag,
  isSet,
  isSetItem,
  isSetList,
  isSetMap,
  isStar,
  isStarIdentifier,
  isString,
  isTable,
  isType,
  isTypeArray,
  isUnaryExpression,
  isUnaryOperator,
  isUpdate,
  isUpdateFrom,
  isUsing,
  isValues,
  isValuesList,
  isWhen,
  isWhere,
  isWithTag,
  isWrappedExpression,
} from './grammar.guards';

export {
  chunk,
  first,
  groupBy,
  identity,
  initial,
  isDiffBy,
  isEmpty,
  isEqual,
  isNil,
  isObject,
  isUnique,
  isUniqueBy,
  last,
  length,
  orderBy,
  tail,
  toMilliseconds,
} from './util';

export { Grammar } from './grammar';

export const parser = Parser<QueryTag | WithTag>(Grammar);
