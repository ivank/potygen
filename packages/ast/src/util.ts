export const length = (item: unknown): number | undefined => (Array.isArray(item) ? item.length : undefined);

export const identity = <T>(item: T): T => item;

export const isEmpty = <T>(items?: T[]): boolean => !items || items?.length === 0;

export function tail<T>(items: [unknown, ...T[]]): T[];
export function tail<T>(items: T[]): T[];
export function tail<T>(items: T[]): T[] {
  return items.slice(1);
}

export function last<T>(items?: [...unknown[], T]): T;
export function last<T>(items?: T[]): T | undefined;
export function last<T>(items?: T[]): T | undefined {
  return items?.[items.length - 1];
}

export function first<T>(items?: [T, ...unknown[]]): T;
export function first<T>(items?: T[]): T | undefined;
export function first<T>(items?: T[]): T | undefined {
  return items?.[0];
}

export function initial<T>(items: [...T[], unknown]): T[];
export function initial<T>(items: T[]): T[];
export function initial<T>(items: T[]): T[] {
  return items.slice(0, -1);
}

export const isUnique =
  <T>(predicate: (item: T) => unknown = identity) =>
  (item: T, index: number, items: T[]): boolean =>
    items.findIndex((current) => predicate(item) === predicate(current)) === index;

export const isUniqueBy =
  <T>(predicate: (a: T, b: T) => boolean = isEqual) =>
  (item: T, index: number, items: T[]): boolean =>
    items.findIndex((current) => predicate(item, current)) === index;

export const orderBy =
  <T>(predicate: (item: T) => any = identity) =>
  (a: T, b: T): number => {
    const orderA = predicate(a);
    const orderB = predicate(b);
    return orderA === orderB ? 0 : orderA > orderB ? 1 : -1;
  };

export const isNil = <T>(item: T | undefined | null): item is T => Boolean(item);

export const isDiffBy =
  <T>(predicate: (item: T) => unknown, to: T[]) =>
  (fromItem: T): boolean =>
    !to.some((toItem) => predicate(fromItem) === predicate(toItem));

export const groupBy = <T, K extends string | number>(predicate: (item: T) => K, items: T[]): Record<K, T[]> =>
  items.reduce((acc, item) => {
    const key = predicate(item);
    return { ...acc, [key]: (acc[key] ?? ([] as T[])).concat(item) };
  }, {} as Record<K, T[]>);

export const chunk = <T>(size: number, items: T[]): T[][] =>
  [...Array(Math.ceil(items.length / size))].map((_, index) => items.slice(index * size, (index + 1) * size));

export const isObject = (item: unknown): item is Record<string, unknown> => typeof item === 'object' && item !== null;

export const isEqual = (a: unknown, b: unknown): boolean => {
  if (a === b) {
    return true;
  }
  if (a === undefined || b === undefined) {
    return false;
  }
  if (typeof a === 'object' && typeof b === 'object') {
    if (a === null || b === null) {
      return false;
    } else if (a instanceof Date && b instanceof Date) {
      return a.getTime() === b.getTime();
    } else if (Array.isArray(a) && Array.isArray(b)) {
      return a.length === b.length && a.every((item, index) => isEqual(item, b[index]));
    } else if (isObject(a) && isObject(b)) {
      const aKeys = Object.keys(a).sort();
      const bKeys = Object.keys(a).sort();
      return isEqual(aKeys, bKeys) && aKeys.every((key) => isEqual(a[key], b[key]));
    }
  }
  return false;
};
