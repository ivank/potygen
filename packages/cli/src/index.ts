export {
  DataTable,
  DataFunction,
  DataEnum,
  DataView,
  DataComposite,
  LoadedDataTable,
  LoadedDataComposite,
  DataViewRaw,
  DataViewParsed,
  LoadedDataView,
  LoadedDataFunction,
  LoadedDataEnum,
  Data,
  LoadedData,
  LoadedParam,
  LoadedResult,
  LoadedQueryInterface,
  LoadedFunction,
  LoadedComposite,
  LoadedSourceTable,
  LoadedSourceQuery,
  LoadedSourceView,
  LoadedSourceValues,
  LoadedSource,
  LoadedContext,
  TemplateTagQuery,
  ParsedTypescriptFile,
  ParsedSqlFile,
  ParsedFile,
  LoadedTemplateTagQuery,
  LoadedTypescriptFile,
  LoadedSqlFile,
  LoadedFile,
  Logger,
  LoadContext,
  QualifiedName,
} from './types';

export {
  isDataTable,
  isDataFunction,
  isDataEnum,
  isLoadedDataTable,
  isLoadedDataFunction,
  isLoadedDataEnum,
  isLoadedDataView,
  isLoadedDataComposite,
} from './guards';

export { toLoadedContext, loadQueryInterfacesData, toLoadedQueryInterface, loadData, extractDataSources } from './load';

export { Config, ConfigType, FullConfigType, toConfig } from './config';

export { emitLoadedFile, toTypeSource } from './emit';
export { glob } from './glob';
