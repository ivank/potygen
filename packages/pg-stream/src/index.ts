export { toInternalCursorGenerator } from './internal-cursor-generator';
export { ToEachBatch, ToEachBatchOptions, toEachBatch } from './each-batch';
export { ToReadable, ToReadableOptions, toReadable } from './readable';
export {
  ToAsyncIterator,
  toAsyncIterator,
  ToIteratorOptions,
  ToAsyncBatchIterator,
  toAsyncBatchIterator,
} from './iterator';
