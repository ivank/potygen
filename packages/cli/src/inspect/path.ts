import { Tag, isNode, SqlTag } from '@potygen/ast';
import { PathItem, Path } from '../types';

const isItem =
  <T extends Tag>(predicate: (tag: SqlTag) => tag is T) =>
  (item: PathItem<Tag>): item is PathItem<T> =>
    predicate(item.tag);

export const closestParent =
  <T extends Tag>(predicate: (tag: SqlTag) => tag is T) =>
  (path: Path): T | undefined =>
    path.find(isItem(predicate))?.tag;

export const closestParentPath =
  <T extends Tag>(predicate: (tag: SqlTag) => tag is T) =>
  (path: Path): PathItem<T> | undefined =>
    path.find(isItem(predicate));

export const toPath = (tag: Tag, offset: number, path: Path = []): Path | undefined => {
  if (isNode(tag)) {
    const index = (tag.values as Tag[]).findIndex((item) => offset >= item.start && offset <= item.end + 1);

    if (index !== -1) {
      return toPath(tag.values[index], offset, [{ tag, index }, ...path]);
    } else {
      return [{ tag }, ...path];
    }
  }
  return [{ tag }, ...path];
};
