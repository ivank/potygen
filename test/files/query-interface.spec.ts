import { Parser, ParserError } from '@ikerin/rd-parse';
import { SqlGrammar } from '../../src/sql.grammar';
import { inspect } from 'util';
import { readdirSync, readFileSync } from 'fs';
import { join } from 'path';
import { convertTag } from '../../src/query-interface';

const sqlParser = Parser(SqlGrammar);

describe('Query Interface', () => {
  it.each(
    readdirSync(join(__dirname, 'sql')).map((filename) => [
      filename,
      readFileSync(join(__dirname, 'sql', filename), 'utf-8'),
    ]),
  )('Should convert complex sql %s', (name, sql) => {
    try {
      expect(convertTag(sqlParser(sql))).toMatchSnapshot(name);
    } catch (e) {
      if (e instanceof ParserError) {
        console.log(inspect(e, { depth: 15, colors: true }));
      }
      throw e;
    }
  });
});
