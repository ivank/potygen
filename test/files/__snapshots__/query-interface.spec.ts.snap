// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query Interface Should convert complex sql account-levelisations.query.sql: account-levelisations.query.sql 1`] = `
Object {
  "params": Array [
    Object {
      "lastPos": 1661,
      "name": "q",
      "pick": Array [],
      "pos": 1605,
      "required": false,
      "spread": false,
      "type": "string",
    },
    Object {
      "lastPos": 1788,
      "name": "levelisationId",
      "pick": Array [],
      "pos": 1706,
      "required": false,
      "spread": false,
      "type": Object {
        "column": "levelisation_id",
        "schema": "public",
        "table": "account_levelisations",
        "type": "column",
      },
    },
    Object {
      "lastPos": 1910,
      "name": "ids",
      "pick": Array [],
      "pos": 1805,
      "required": false,
      "spread": false,
      "type": Object {
        "items": "number",
        "type": "array",
      },
    },
    Object {
      "lastPos": 1942,
      "name": "sortField",
      "pick": Array [],
      "pos": 1931,
      "required": false,
      "spread": false,
      "type": "string",
    },
    Object {
      "lastPos": 1970,
      "name": "sortOrder",
      "pick": Array [],
      "pos": 1959,
      "required": false,
      "spread": false,
      "type": "string",
    },
    Object {
      "lastPos": 2581,
      "name": "limit",
      "pick": Array [],
      "pos": 2575,
      "required": false,
      "spread": false,
      "type": "number",
    },
    Object {
      "lastPos": 2601,
      "name": "offset",
      "pick": Array [],
      "pos": 2594,
      "required": false,
      "spread": false,
      "type": "number",
    },
  ],
  "result": Array [
    Object {
      "name": "id",
      "type": Object {
        "column": "id",
        "schema": "public",
        "table": "account_levelisations",
        "type": "column",
      },
    },
    Object {
      "name": "levelisationId",
      "type": Object {
        "column": "levelisation_id",
        "schema": "public",
        "table": "account_levelisations",
        "type": "column",
      },
    },
    Object {
      "name": "accountId",
      "type": Object {
        "column": "account_id",
        "schema": "public",
        "table": "account_levelisations",
        "type": "column",
      },
    },
    Object {
      "name": "installationId",
      "type": Object {
        "column": "installation_id",
        "schema": "public",
        "table": "account_levelisations",
        "type": "column",
      },
    },
    Object {
      "name": "state",
      "type": Object {
        "column": "state",
        "schema": "public",
        "table": "account_levelisations",
        "type": "column",
      },
    },
    Object {
      "name": "isAccepted",
      "type": Object {
        "column": "is_accepted",
        "schema": "public",
        "table": "account_levelisations",
        "type": "column",
      },
    },
    Object {
      "name": "generationStartReadOn",
      "type": Object {
        "column": "generation_start_read_on",
        "schema": "public",
        "table": "account_levelisations",
        "type": "column",
      },
    },
    Object {
      "name": "generationStartReadValue",
      "type": Object {
        "column": "generation_start_read_value",
        "schema": "public",
        "table": "account_levelisations",
        "type": "column",
      },
    },
    Object {
      "name": "generationEndReadOn",
      "type": Object {
        "column": "generation_end_read_on",
        "schema": "public",
        "table": "account_levelisations",
        "type": "column",
      },
    },
    Object {
      "name": "generationEndReadValue",
      "type": Object {
        "column": "generation_end_read_value",
        "schema": "public",
        "table": "account_levelisations",
        "type": "column",
      },
    },
    Object {
      "name": "generationPercentageSplit",
      "type": Object {
        "column": "generation_percentage_split",
        "schema": "public",
        "table": "account_levelisations",
        "type": "column",
      },
    },
    Object {
      "name": "generationPayment",
      "type": Object {
        "column": "generation_payment",
        "schema": "public",
        "table": "account_levelisations",
        "type": "column",
      },
    },
    Object {
      "name": "generationEnergy",
      "type": Object {
        "column": "generation_energy",
        "schema": "public",
        "table": "account_levelisations",
        "type": "column",
      },
    },
    Object {
      "name": "exportStartReadOn",
      "type": Object {
        "column": "export_start_read_on",
        "schema": "public",
        "table": "account_levelisations",
        "type": "column",
      },
    },
    Object {
      "name": "exportStartReadValue",
      "type": Object {
        "column": "export_start_read_value",
        "schema": "public",
        "table": "account_levelisations",
        "type": "column",
      },
    },
    Object {
      "name": "exportEndReadOn",
      "type": Object {
        "column": "export_end_read_on",
        "schema": "public",
        "table": "account_levelisations",
        "type": "column",
      },
    },
    Object {
      "name": "exportEndReadValue",
      "type": Object {
        "column": "export_end_read_value",
        "schema": "public",
        "table": "account_levelisations",
        "type": "column",
      },
    },
    Object {
      "name": "exportPercentageSplit",
      "type": Object {
        "column": "export_percentage_split",
        "schema": "public",
        "table": "account_levelisations",
        "type": "column",
      },
    },
    Object {
      "name": "exportPayment",
      "type": Object {
        "column": "export_payment",
        "schema": "public",
        "table": "account_levelisations",
        "type": "column",
      },
    },
    Object {
      "name": "exportEnergy",
      "type": Object {
        "column": "export_energy",
        "schema": "public",
        "table": "account_levelisations",
        "type": "column",
      },
    },
    Object {
      "name": "totalPayment",
      "type": Object {
        "column": "total_payment",
        "schema": "public",
        "table": "account_levelisations",
        "type": "column",
      },
    },
    Object {
      "name": "vatPayment",
      "type": Object {
        "column": "vat_payment",
        "schema": "public",
        "table": "account_levelisations",
        "type": "column",
      },
    },
    Object {
      "name": "cfrFitId",
      "type": Object {
        "column": "scheme_account_reference",
        "schema": "public",
        "table": "contracts",
        "type": "column",
      },
    },
  ],
}
`;

exports[`Query Interface Should convert complex sql account-levelisations-set-bulk-update.query.sql: account-levelisations-set-bulk-update.query.sql 1`] = `
Object {
  "params": Array [
    Object {
      "lastPos": 542,
      "name": "ids",
      "pick": Array [],
      "pos": 529,
      "required": false,
      "spread": false,
      "type": Object {
        "items": "number",
        "type": "array",
      },
    },
  ],
  "result": Array [
    Object {
      "name": "id",
      "type": Object {
        "column": "id",
        "schema": "public",
        "table": "account_levelisations",
        "type": "column",
      },
    },
  ],
}
`;

exports[`Query Interface Should convert complex sql accounts.query.sql: accounts.query.sql 1`] = `
Object {
  "params": Array [
    Object {
      "lastPos": 2255,
      "name": "q",
      "pick": Array [],
      "pos": 1192,
      "required": false,
      "spread": false,
      "type": "string",
    },
    Object {
      "lastPos": 2329,
      "name": "ids",
      "pick": Array [],
      "pos": 2272,
      "required": false,
      "spread": false,
      "type": Object {
        "items": "number",
        "type": "array",
      },
    },
    Object {
      "lastPos": 2634,
      "name": "limit",
      "pick": Array [],
      "pos": 2628,
      "required": false,
      "spread": false,
      "type": "number",
    },
    Object {
      "lastPos": 2654,
      "name": "offset",
      "pick": Array [],
      "pos": 2647,
      "required": false,
      "spread": false,
      "type": "number",
    },
  ],
  "result": Array [
    Object {
      "name": "id",
      "type": Object {
        "column": "id",
        "schema": "public",
        "table": "accounts",
        "type": "column",
      },
    },
    Object {
      "name": "state",
      "type": Object {
        "column": "state",
        "schema": "public",
        "table": "accounts",
        "type": "column",
      },
    },
    Object {
      "name": "startOn",
      "type": Object {
        "column": "start_on",
        "schema": "public",
        "table": "accounts",
        "type": "column",
      },
    },
    Object {
      "name": "firstName",
      "type": Object {
        "column": "first_name",
        "schema": "public",
        "table": "contacts",
        "type": "column",
      },
    },
    Object {
      "name": "lastName",
      "type": Object {
        "column": "last_name",
        "schema": "public",
        "table": "contacts",
        "type": "column",
      },
    },
    Object {
      "name": "email",
      "type": Object {
        "column": "email",
        "schema": "public",
        "table": "contacts",
        "type": "column",
      },
    },
    Object {
      "name": "address",
      "type": Object {
        "args": Array [
          Object {
            "column": "address_line_1",
            "schema": "public",
            "table": "installation_addresses",
            "type": "column",
          },
          Object {
            "column": "address_line_2",
            "schema": "public",
            "table": "installation_addresses",
            "type": "column",
          },
          Object {
            "column": "city",
            "schema": "public",
            "table": "installation_addresses",
            "type": "column",
          },
          Object {
            "column": "postcode",
            "schema": "public",
            "table": "installation_addresses",
            "type": "column",
          },
        ],
        "name": "concat_ws",
        "type": "function",
      },
    },
    Object {
      "name": "fitIds",
      "type": Object {
        "args": Array [
          Object {
            "column": "scheme_account_reference",
            "schema": "public",
            "table": "contracts",
            "type": "column",
          },
        ],
        "name": "array_agg",
        "type": "function",
      },
    },
  ],
}
`;

exports[`Query Interface Should convert complex sql accounts.sql: accounts.sql 1`] = `
Object {
  "params": Array [],
  "result": Array [],
}
`;

exports[`Query Interface Should convert complex sql bacs.query.sql: bacs.query.sql 1`] = `
Object {
  "params": Array [
    Object {
      "lastPos": 620,
      "name": "quarter",
      "pick": Array [],
      "pos": 611,
      "required": false,
      "spread": false,
      "type": Object {
        "column": "levelisation_reference",
        "schema": "public",
        "table": "payments",
        "type": "column",
      },
    },
    Object {
      "lastPos": 635,
      "name": "perPage",
      "pick": Array [],
      "pos": 626,
      "required": false,
      "spread": false,
      "type": "string",
    },
    Object {
      "lastPos": 650,
      "name": "offset",
      "pick": Array [],
      "pos": 642,
      "required": false,
      "spread": false,
      "type": "string",
    },
  ],
  "result": Array [
    Object {
      "name": "fitId",
      "type": Object {
        "column": "fit_reference",
        "schema": "public",
        "table": "payments",
        "type": "column",
      },
    },
    Object {
      "name": "name",
      "type": Object {
        "column": "beneficiary_name",
        "schema": "public",
        "table": "accounts",
        "type": "column",
      },
    },
    Object {
      "name": "amount",
      "type": Object {
        "column": "amount",
        "schema": "public",
        "table": "payments",
        "type": "column",
      },
    },
    Object {
      "name": "quarter",
      "type": Object {
        "column": "levelisation_reference",
        "schema": "public",
        "table": "payments",
        "type": "column",
      },
    },
    Object {
      "name": "bankDetails",
      "type": Object {
        "args": Array [
          Object {
            "column": "beneficiary_sort_code",
            "schema": "public",
            "table": "a",
            "type": "column",
          },
          Object {
            "column": "beneficiary_account_number",
            "schema": "public",
            "table": "a",
            "type": "column",
          },
        ],
        "name": "concat",
        "type": "function",
      },
    },
    Object {
      "name": "addressLine1",
      "type": Object {
        "column": "address_line_1",
        "schema": "public",
        "table": "addresses",
        "type": "column",
      },
    },
    Object {
      "name": "addressLine2",
      "type": Object {
        "column": "address_line_2",
        "schema": "public",
        "table": "addresses",
        "type": "column",
      },
    },
    Object {
      "name": "addressLine3",
      "type": Object {
        "column": "address_line_3",
        "schema": "public",
        "table": "addresses",
        "type": "column",
      },
    },
    Object {
      "name": "postcode",
      "type": Object {
        "column": "postcode",
        "schema": "public",
        "table": "addresses",
        "type": "column",
      },
    },
  ],
}
`;

exports[`Query Interface Should convert complex sql contracts.sql: contracts.sql 1`] = `
Object {
  "params": Array [],
  "result": Array [],
}
`;

exports[`Query Interface Should convert complex sql delete-accounts.query.sql: delete-accounts.query.sql 1`] = `
Object {
  "params": Array [
    Object {
      "lastPos": 263,
      "name": "type",
      "pick": Array [],
      "pos": 251,
      "required": false,
      "spread": false,
      "type": Object {
        "column": "scheme_type",
        "schema": "public",
        "table": "contracts",
        "type": "column",
      },
    },
    Object {
      "lastPos": 291,
      "name": "state",
      "pick": Array [],
      "pos": 284,
      "required": false,
      "spread": false,
      "type": Object {
        "column": "state",
        "schema": "public",
        "table": "accounts",
        "type": "column",
      },
    },
  ],
  "result": Array [
    Object {
      "name": "id",
      "type": Object {
        "column": "id",
        "schema": "public",
        "table": "accounts",
        "type": "column",
      },
    },
    Object {
      "name": "sourceSystemId",
      "type": Object {
        "column": "source_system_id",
        "schema": "public",
        "table": "accounts",
        "type": "column",
      },
    },
    Object {
      "name": "customerId",
      "type": Object {
        "column": "customer_id",
        "schema": "public",
        "table": "accounts",
        "type": "column",
      },
    },
  ],
}
`;

exports[`Query Interface Should convert complex sql generation-read-data.query.sql: generation-read-data.query.sql 1`] = `
Object {
  "params": Array [
    Object {
      "lastPos": 223,
      "name": "installationId",
      "pick": Array [],
      "pos": 205,
      "required": false,
      "spread": false,
      "type": Object {
        "column": "installation_id",
        "schema": "public",
        "table": "installation_meters",
        "type": "column",
      },
    },
    Object {
      "lastPos": 454,
      "name": "exportDateOn",
      "pick": Array [],
      "pos": 429,
      "required": false,
      "spread": false,
      "type": "Date",
    },
  ],
  "result": Array [
    Object {
      "name": "dateOn",
      "type": Object {
        "column": "date_on",
        "schema": "public",
        "table": "meter_reads",
        "type": "column",
      },
    },
    Object {
      "name": "value",
      "type": Object {
        "column": "value",
        "schema": "public",
        "table": "meter_reads",
        "type": "column",
      },
    },
  ],
}
`;

exports[`Query Interface Should convert complex sql installation-meter-update.query.sql: installation-meter-update.query.sql 1`] = `
Object {
  "params": Array [
    Object {
      "lastPos": 98,
      "name": "meterId",
      "pick": Array [],
      "pos": 89,
      "required": false,
      "spread": false,
      "type": Object {
        "column": "meter_id",
        "schema": "public",
        "table": "installation_meters",
        "type": "column",
      },
    },
    Object {
      "lastPos": 53,
      "name": "removalDate",
      "pick": Array [],
      "pos": 40,
      "required": false,
      "spread": false,
      "type": "unknown",
    },
  ],
  "result": Array [],
}
`;

exports[`Query Interface Should convert complex sql installations.sql: installations.sql 1`] = `
Object {
  "params": Array [],
  "result": Array [],
}
`;

exports[`Query Interface Should convert complex sql meter-reads.sql: meter-reads.sql 1`] = `
Object {
  "params": Array [],
  "result": Array [],
}
`;

exports[`Query Interface Should convert complex sql rates-by-tarrif-ids.query.sql: rates-by-tarrif-ids.query.sql 1`] = `
Object {
  "params": Array [
    Object {
      "lastPos": 320,
      "name": "ids",
      "pick": Array [],
      "pos": 314,
      "required": false,
      "spread": true,
      "type": Object {
        "column": "id",
        "schema": "public",
        "table": "tariffs",
        "type": "column",
      },
    },
  ],
  "result": Array [
    Object {
      "name": "tariffId",
      "type": Object {
        "column": "tariff_id",
        "schema": "public",
        "table": "tariff_rates",
        "type": "column",
      },
    },
    Object {
      "name": "rate",
      "type": Object {
        "column": "rate",
        "schema": "public",
        "table": "tariff_rates",
        "type": "column",
      },
    },
    Object {
      "name": "startOn",
      "type": Object {
        "column": "start_date_on",
        "schema": "public",
        "table": "tariff_rates",
        "type": "column",
      },
    },
    Object {
      "name": "tariffCode",
      "type": Object {
        "column": "code",
        "schema": "public",
        "table": "tariffs",
        "type": "column",
      },
    },
    Object {
      "name": "tariffType",
      "type": Object {
        "column": "type",
        "schema": "public",
        "table": "tariffs",
        "type": "column",
      },
    },
    Object {
      "name": "endOn",
      "type": Object {
        "column": "end_date_on",
        "schema": "public",
        "table": "tariff_rates",
        "type": "column",
      },
    },
  ],
}
`;

exports[`Query Interface Should convert complex sql read-create.query.sql: read-create.query.sql 1`] = `
Object {
  "params": Array [
    Object {
      "lastPos": 215,
      "name": "meterId",
      "pick": Array [],
      "pos": 92,
      "required": false,
      "spread": false,
      "type": Object {
        "column": "meter_id",
        "schema": "public",
        "table": "meter_reads",
        "type": "column",
      },
    },
    Object {
      "lastPos": 215,
      "name": "reason",
      "pick": Array [],
      "pos": 102,
      "required": false,
      "spread": false,
      "type": Object {
        "column": "reason",
        "schema": "public",
        "table": "meter_reads",
        "type": "column",
      },
    },
    Object {
      "lastPos": 215,
      "name": "type",
      "pick": Array [],
      "pos": 111,
      "required": false,
      "spread": false,
      "type": Object {
        "column": "type",
        "schema": "public",
        "table": "meter_reads",
        "type": "column",
      },
    },
    Object {
      "lastPos": 215,
      "name": "dateOn",
      "pick": Array [],
      "pos": 127,
      "required": false,
      "spread": false,
      "type": Object {
        "column": "date_on",
        "schema": "public",
        "table": "meter_reads",
        "type": "column",
      },
    },
    Object {
      "lastPos": 215,
      "name": "submittedAt",
      "pick": Array [],
      "pos": 160,
      "required": false,
      "spread": false,
      "type": Object {
        "column": "submitted_at",
        "schema": "public",
        "table": "meter_reads",
        "type": "column",
      },
    },
    Object {
      "lastPos": 215,
      "name": "value",
      "pick": Array [],
      "pos": 198,
      "required": false,
      "spread": false,
      "type": "number",
    },
  ],
  "result": Array [
    Object {
      "name": "id",
      "type": Object {
        "column": "id",
        "schema": "public",
        "table": "meter_reads",
        "type": "column",
      },
    },
    Object {
      "name": "value",
      "type": Object {
        "column": "value",
        "schema": "public",
        "table": "meter_reads",
        "type": "column",
      },
    },
    Object {
      "name": "meterId",
      "type": Object {
        "column": "meter_id",
        "schema": "public",
        "table": "meter_reads",
        "type": "column",
      },
    },
    Object {
      "name": "submittedAt",
      "type": Object {
        "column": "submitted_at",
        "schema": "public",
        "table": "meter_reads",
        "type": "column",
      },
    },
  ],
}
`;

exports[`Query Interface Should convert complex sql reads-by-account-id.query.sql: reads-by-account-id.query.sql 1`] = `
Object {
  "params": Array [
    Object {
      "lastPos": 597,
      "name": "id",
      "pick": Array [],
      "pos": 593,
      "required": false,
      "spread": false,
      "type": Object {
        "column": "account_id",
        "schema": "public",
        "table": "contracts",
        "type": "column",
      },
    },
    Object {
      "lastPos": 632,
      "name": "offset",
      "pick": Array [],
      "pos": 624,
      "required": false,
      "spread": false,
      "type": "string",
    },
    Object {
      "lastPos": 647,
      "name": "perPage",
      "pick": Array [],
      "pos": 638,
      "required": false,
      "spread": false,
      "type": "string",
    },
  ],
  "result": Array [
    Object {
      "name": "id",
      "type": Object {
        "column": "id",
        "schema": "public",
        "table": "meter_reads",
        "type": "column",
      },
    },
    Object {
      "name": "msn",
      "type": Object {
        "column": "serial_number",
        "schema": "public",
        "table": "meters",
        "type": "column",
      },
    },
    Object {
      "name": "meterId",
      "type": Object {
        "column": "meter_id",
        "schema": "public",
        "table": "meter_reads",
        "type": "column",
      },
    },
    Object {
      "name": "dateOn",
      "type": Object {
        "column": "date_on",
        "schema": "public",
        "table": "meter_reads",
        "type": "column",
      },
    },
    Object {
      "name": "value",
      "type": Object {
        "column": "value",
        "schema": "public",
        "table": "meter_reads",
        "type": "column",
      },
    },
    Object {
      "name": "type",
      "type": Object {
        "column": "type",
        "schema": "public",
        "table": "meter_reads",
        "type": "column",
      },
    },
    Object {
      "name": "reason",
      "type": Object {
        "column": "reason",
        "schema": "public",
        "table": "meter_reads",
        "type": "column",
      },
    },
    Object {
      "name": "createdAt",
      "type": Object {
        "column": "created_at",
        "schema": "public",
        "table": "meter_reads",
        "type": "column",
      },
    },
    Object {
      "name": "updatedAt",
      "type": Object {
        "column": "updated_at",
        "schema": "public",
        "table": "meter_reads",
        "type": "column",
      },
    },
    Object {
      "name": "sourceSystemId",
      "type": Object {
        "column": "source_system_id",
        "schema": "public",
        "table": "meter_reads",
        "type": "column",
      },
    },
    Object {
      "name": "mpan",
      "type": Object {
        "column": "mpan",
        "schema": "public",
        "table": "meters",
        "type": "column",
      },
    },
  ],
}
`;

exports[`Query Interface Should convert complex sql tariff-rates.sql: tariff-rates.sql 1`] = `
Object {
  "params": Array [],
  "result": Array [],
}
`;
